File: paper-code-with-ballot/Derivation.hs
{-# LANGUAGE StandaloneDeriving, FlexibleInstances #-}
module Derivation where
import qualified Prelude as P
import Lib

c2hl :: List a -> [a]
c2hl Nil = []
c2hl (Cons x xs) = x : (c2hl xs)

deriving instance (P.Eq Cand)
deriving instance (P.Ord Cand)
deriving instance (P.Show Cand)
deriving instance (P.Show Bool)
deriving instance (P.Eq Bool)
deriving instance (P.Show Nat)
deriving instance (P.Eq Nat)
deriving instance (P.Ord Nat)
deriving instance (P.Show a, P.Show b) => P.Show (Sum a b)
deriving instance (P.Show a, P.Show b) => P.Show (Prod a b)

deriving instance (P.Show a) => P.Show (List a)
deriving instance (P.Show Comparison)
deriving instance (P.Show Sumbool)
deriving instance (P.Show a) => P.Show (Sumor a)
deriving instance (P.Show Positive)
deriving instance (P.Show Z)

haskInt :: Nat -> P.Int
haskInt O = 0
haskInt (S p) = 1 P.+ haskInt p

haskPos :: Positive -> P.Int
haskPos XH = 1
haskPos p = haskHelp p 1 where
  haskHelp XH p2 = p2
  haskHelp (XO r) p2 = haskHelp r (2 P.* p2)
  haskHelp (XI r) p2 = p2 P.+ haskHelp r (2 P.* p2)


haskZ :: Z -> P.Int
haskZ Z0 = 0
haskZ (Zpos p) = haskPos p
haskZ (Zneg p) = -1 P.* haskPos p

instance P.Show PathT where
  show (UnitT x y) = P.show x P.++ " --> " P.++ P.show y
  show (ConsT x _ _ p) = P.show x P.++ " --> " P.++ P.show p



-- deriving instance (P.Show PathT)
instance P.Show (SigT (Cand -> Bool) (SigT Count ())) where
  show (ExistT f (ExistT v _)) = P.show v


show_winner :: Wins_type -> Cand -> [Cand] -> P.String
show_winner g x [] = ""
show_winner g x (y : ys) =
  case (g y) of 
   (ExistT u (Pair v (ExistT f _))) -> 
    "   for " P.++ P.show y P.++ ": " P.++ "path " P.++ P.show v P.++ " of strenght "  P.++ P.show (haskZ u) P.++  ", " P.++ 
    P.show (1 P.+ haskZ u) P.++ "-" P.++ "coclosed set: " P.++ P.show (P.filter (\(xx, yy) -> f (Pair xx yy) P.== True) [(a, b) | a <- (c2hl cand_all), b <- (c2hl cand_all), a P./= b])
    P.++ "\n" P.++ show_winner g x ys


show_loser :: Loses_type -> Cand -> P.String
show_loser g x = 
  case g of 
   (ExistT u (ExistT c (Pair p (ExistT f _)))) -> 
    "   for " P.++ P.show c P.++ ": " P.++ "path " P.++ P.show p P.++ " of strength >= " P.++ P.show (haskZ u) P.++ ", " P.++
    P.show (haskZ u) P.++ "-" P.++ "coclosed set: " P.++ P.show (P.filter (\(xx, yy) -> f (Pair xx yy) P.== True) [(a, b) | a <- (c2hl cand_all), b <- (c2hl cand_all), a P./= b])


show_cand :: (Cand -> Sum Wins_type Loses_type) -> Cand -> P.String
show_cand f x =
   case (f x) of
    Inl g -> "winning: " P.++ P.show x  P.++ "\n" P.++ show_winner g x (P.filter (\y -> y P./= x) (c2hl cand_all))
    Inr h -> "losing: "  P.++ P.show x  P.++ "\n" P.++ show_loser h x

show_ballot :: Ballot -> P.String
show_ballot f = P.unwords (P.map (\x -> P.show x P.++ P.show (haskInt (f x)))  (c2hl cand_all))

show_list_ballot :: List Ballot -> P.String
show_list_ballot ls = P.show (P.map show_ballot (c2hl ls))

show_marg :: (Cand -> Cand -> Z) -> P.String
show_marg m = "[" P.++ P.unwords (P.map (\(x, y) -> P.show x P.++ P.show y P.++ ":" P.++ P.show (haskZ (m x y))) [(a, b) | a <- (c2hl cand_all), b <- (c2hl cand_all), b P.> a]) P.++ "]"

bool_b :: List Ballot -> P.Bool
bool_b Nil = P.True
bool_b _ = P.False

instance P.Show Count where
  show (Ax ls m) = ""
  show (Cvalid u us m nm inbs c) = P.show c P.++ "V = [" P.++ show_ballot u P.++ (if bool_b us P.== P.True then "]" else ",....]") P.++ ", I = " P.++ show_list_ballot inbs  P.++ ", M = " P.++ show_marg m
                                      P.++ "\n----------------------------------------------------------------------------------------------------------------------------------------------------\n"
  show (Cinvalid u us m inbs c) = P.show c P.++ "I = [" P.++ show_ballot u P.++ (if bool_b us P.== P.True then "]" else ",....]") P.++ ", I = " P.++ show_list_ballot inbs P.++ ", M = " P.++ show_marg m
                                      P.++ "\n----------------------------------------------------------------------------------------------------------------------------------------------------\n"
  show (Fin m ls p f c) =  P.show c P.++ "V = [], I = " P.++ show_list_ballot ls P.++ ", M = " P.++ show_marg m
                                    P.++ "\n----------------------------------------------------------------------------------------------------------------------------------------------------\n"
                                    P.++ (P.unlines P.$ P.map (show_cand f) (c2hl cand_all))

File: SchulzeOCaml/derivation.ml
open Lib
open Printf
       
let show_bool = function 
  | Lib.True -> "True"
  | Lib.False -> "False"

let rec nat_int = function
  | O -> 0
  | S n' -> 1 + nat_int n'

let rec ocamllist = function
  | Nil -> []
  | Cons (h, t) -> h :: ocamllist t
                                  
let show_nat n = string_of_int (nat_int n)         

let rec ocaml_pos = function
  | Lib.XH -> 1
  | p ->
     let rec ocaml_help t c =
       match t with
       | Lib.XH -> c
       | Lib.XO r -> ocaml_help r (2 * c)
       | Lib.XI r -> c + ocaml_help r (2 * c)
     in ocaml_help p 1

let ocaml_z = function
  | Lib.Z0 -> 0
  | Lib.Zpos p -> ocaml_pos p
  | Lib.Zneg p -> -1 * ocaml_pos p

let show_z p = string_of_int (ocaml_z p)
                             
let show_cand = function
  | Lib.A -> "A"
  | Lib.B -> "B"
  | Lib.C -> "C"
  | Lib.D -> "D"

let compare x y =
  match x, y with
  | Lib.A, Lib.A | Lib.B, Lib.B | Lib.C, Lib.C | Lib.D, Lib.D -> true
  | _, _ -> false

let compare_pair x y =
  match x, y with
  | Lib.Pair (x1, y1), Lib.Pair (x2, y2) -> compare x1 x2 && compare y1 y2
                                                                     
let rec cross_prod_orig = function
  | [] -> []
  | h :: t as v -> List.map (fun x -> (h, x)) v @ cross_prod_orig t
                                                                  
let show_cand_pair (c1, c2) = "(" ^ show_cand c1 ^ "," ^ show_cand c2 ^ ")"

let compare_truth x y =
  match x, y with
  | Lib.True, Lib.True | Lib.False, Lib.False -> true
  | _, _ -> false
  
let show_coclosed f l =
  "[" ^ String.concat
          ","
          (List.map show_cand_pair  
          (List.filter (fun (x, y) -> compare_truth (f (Lib.Pair (x, y))) Lib.True)
                       (cross_prod_orig l))) ^ "]"
                          
let show_help f c = show_cand c ^ show_nat (f c)
  
let show_ballot f =
  String.concat " " (List.map (fun c -> show_help f c) (ocamllist Lib.cand_all))
                
let bool_b = function
  | Nil -> true
  | _ -> false

let show_list_inv_ballot = function
  | Nil ->"[]"
  | Cons (b, Nil) -> "[" ^ show_ballot b ^ "]"
  | Cons (b, _) -> "[" ^ show_ballot b ^ ",..]"

let rec cross_prod = function
  | [] -> []
  | h :: t -> List.map (fun x -> (h, x)) t @ cross_prod t
                                                     
let show_pair c1 c2 n = show_cand c1 ^ show_cand c2 ^ ":" ^ show_z n                                           
let show_marg m =
  "[" ^ String.concat
          " "
          (List.map (fun (x, y) -> show_pair x y (m x y)) (cross_prod (ocamllist Lib.cand_all)))
  ^ "]"
      
let rec show_path = function
  | Lib.UnitT (x, y) -> show_cand x ^ " --> " ^ show_cand y
  | Lib.ConsT (x, _, _, p) -> show_cand x ^ " --> " ^ show_path p

let rec show_winner g x xs =
  match xs with
  | [] -> ""
  | y :: ys -> if compare x y then show_winner g x ys
               else
                 match (g y) with
                 | ExistT (u, Pair (v, ExistT (f, _))) ->
                    "   for " ^ show_cand y ^ ": " ^ "path " ^
                      show_path v ^ " of strenght "  ^ show_z u ^  ", " ^
                        string_of_int (1 + ocaml_z u) ^ 
                          "-" ^ "coclosed set: " ^ show_coclosed f (ocamllist Lib.cand_all)
                       ^ (if ys = [] then " " else "\n")  ^ show_winner g x ys 
                         
let show_loser g x =
  match g with
  | ExistT (u, ExistT (c, Pair (p, ExistT (f, _)))) ->
     "   exists " ^ show_cand c ^ ": " ^ "path " ^ show_path p ^ " of strength "
     ^ show_z u ^ ", " ^ show_z u ^ "-" ^ "coclosed set: " ^ show_coclosed f (ocamllist Lib.cand_all)
      
let show_cand f x =
  match f x with
  | Inl g -> "winning: " ^ show_cand x ^ "\n" ^ show_winner g x (ocamllist Lib.cand_all)
  | Inr h -> "losing: " ^ show_cand x ^ "\n" ^ show_loser h x
                                                          
let rec add_string = function
  | 0 -> ""
  | n -> "-" ^ add_string (n - 1)
                            
let underline s = s ^ "\n" ^ add_string (String.length s) ^ "\n"

let rec show_count = function
  | Lib.Ax (_, _) -> ""
  | Lib.Cvalid (u, us, m, nm, inbs, c) ->
     show_count c ^ underline (
       "V: [" ^ show_ballot u ^ (if bool_b us then "]" else ",..]")  ^
         ", I:  " ^ show_list_inv_ballot inbs  ^ ", M: " ^ show_marg m ) 
  | Lib.Cinvalid (u, us, m, inbs, c)  ->
     show_count c ^ underline (
       "V: [" ^ show_ballot u ^
         (if bool_b us then "]" else ",..]") ^ ", I: " ^ show_list_inv_ballot inbs ^ 
           ", M: " ^ show_marg m )
  | Lib.Fin (m, ls, p, f, c) ->
     show_count c ^ underline (
       "V: [], I: " ^ show_list_inv_ballot ls ^ ", M: " ^ show_marg m )
     ^ String.concat "\n" (List.map (fun x -> show_cand f x) (ocamllist Lib.cand_all))
     ^ "\n"

